crates/pipeline_manager/src/local_runner.rs

```diff
@@ -0,0 +1,771 @@
+    let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
+    tokio::spawn(crate::db_notifier::listen(db.clone(), tx));
+    loop {
+        trace!("Waiting for notification");
+        if let Some(DbNotification::Pipeline(op, tenant_id, pipeline_id)) = rx.recv().await {
+            trace!("Received DbNotification {op:?} {tenant_id} {pipeline_id}");
+            pipelines
+                .lock()
+                .await
+                .entry(pipeline_id)
+                .or_insert_with(|| {
+                    let notifier = Arc::new(Notify::new());
+                    spawn(
+                        PipelineAutomaton::new(
+                            pipeline_id,
+                            tenant_id,
+                            &config,
+                            db.clone(),
+                            notifier.clone(),
+                        )
+                        .run(),
+                    );
+                    notifier
+                })
+                .notify_one();
+        }
+    }
+}
+
+pub async fn binary_ref_to_url(
+    binary_ref: &str,
+    pipeline_id: PipelineId,
+) -> Result<String, RunnerError> {
+    let parsed =
+        url::Url::parse(binary_ref).expect("Can only be invoked with valid URLs created by us");
+    match parsed.scheme() {
+        // A file scheme assumes the binary is available locally where
+        // the runner is located.
+        "file" => {
+            let exists = fs::try_exists(parsed.path()).await;
+            match exists {
+                Ok(true) => Ok(parsed.path().to_string()),
+                Ok(false) => Err(RunnerError::BinaryFetchError {
+                    pipeline_id,
+                    error: format!(
+                        "fileref {} for binary required by pipeline {pipeline_id} does not exist",
+                        parsed.path()
+                    ),
+                }),
+                Err(e) => Err(RunnerError::BinaryFetchError {
+                    pipeline_id,
+                    error: format!(
+                        "Check fileref {} for binary required by pipeline {pipeline_id} returned an error: {}",
+                        parsed.path(),
+                        e
+                    ),
+                }),
+            }
+        }
+        _ => todo!("Unsupported URL scheme for binary ref"),
+    }
+}
+
+/// Send HTTP request to pipeline and parse response as a JSON object.
+async fn pipeline_http_request_json_response(
+    pipeline_id: PipelineId,
+    method: Method,
+    endpoint: &str,
+    port: &str,
+) -> Result<(StatusCode, JsonValue), RunnerError> {
+    let response = RunnerApi::pipeline_http_request(pipeline_id, method, endpoint, port).await?;
+    let status = response.status();
+
+    let value = response
+        .json::<JsonValue>()
+        .await
+        .map_err(|e| RunnerError::HttpForwardError {
+            pipeline_id,
+            error: e.to_string(),
+        })?;
+
+    Ok((status, value))
+}
```

Comment:
seems like just some moving around of code I already saw at some point so I mostly ignore this file

Reply:
Yup, this file moves the local-runner code from runner.rs, but switches to using the `db_notifier` instead of being poked directly by the manager.